src/gateway/bridge/shard_queuer.rs

```diff
@@ -102,35 +100,54 @@ impl ShardQueuer {
     /// **Note**: This should be run in its own thread due to the blocking nature of the loop.
     #[instrument(skip(self))]
     pub async fn run(&mut self) {
-        // The duration to timeout from reads over the Rx channel. This can be done in a loop, and
-        // if the read times out then a shard can be started if one is presently waiting in the
-        // queue.
+        // We read from the Rx channel in a loop, and use a timeout of 5 seconds so that we don't
+        // hang forever. When we receive a command to start a shard, we append it to our queue. The
+        // queue is popped in batches of shards, which are started in parallel. A batch is fired
+        // every 5 seconds at minimum in order to avoid being ratelimited.
         const TIMEOUT: Duration = Duration::from_secs(WAIT_BETWEEN_BOOTS_IN_SECONDS);
 
         loop {
-            match timeout(TIMEOUT, self.rx.next()).await {
-                Ok(Some(ShardQueuerMessage::Shutdown)) => {
-                    debug!("[Shard Queuer] Received to shutdown.");
-                    self.shutdown_runners().await;
-
-                    break;
-                },
-                Ok(Some(ShardQueuerMessage::ShutdownShard(shard, code))) => {
-                    debug!("[Shard Queuer] Received to shutdown shard {} with {}.", shard.0, code);
-                    self.shutdown(shard, code).await;
-                },
-                Ok(Some(ShardQueuerMessage::Start(shard_id))) => {
-                    self.checked_start(shard_id).await;
-                },
-                Ok(Some(ShardQueuerMessage::SetShardTotal(shard_total))) => {
-                    self.shard_total = shard_total;
-                },
-                Ok(None) => break,
-                Err(_) => {
-                    if let Some(shard) = self.queue.pop_front() {
-                        self.checked_start(shard).await;
-                    }
-                },
+            if let Ok(msg) = timeout(TIMEOUT, self.rx.next()).await {
+                match msg {
+                    Some(ShardQueuerMessage::SetShardTotal(shard_total)) => {
+                        self.shard_total = shard_total;
+                    },
+                    Some(ShardQueuerMessage::Start(shard_id, concurrent)) => {
+                        if concurrent {
```

Comment:
Non blocker: Can this concurrent bool be part of the pattern, to flatten indentation a bit.

Reply:
Do you mean splitting the match into two separate arms with guards, one for `if concurrent` and one for `if !concurrent`? That seems less readable, and also doesn't satisfy the compiler for exhaustive matching.

Reply:
I do mean true or false literals.

Reply:
Unfortunately this won't play that well with struct-style variants, because rustfmt wants to indent the fields in the match arm, so you end up with something like:
```rust
Some(ShardQueuerMessage::Start {
    shard_id,
    concurrent: true,
}) => {
    /// ...
},
Some(ShardQueuerMessage::Start {
    shard_id,
    concurrent: false,
}) => {
    /// ...
},
```
As opposed to:
```rust
Some(ShardQueuerMessage::Start {
    shard_id,
    concurrent,
}) => {
    if concurrent {
        /// ...
    } else {
        /// ...
    }
},

Reply:
Oh that sucks, pick your poison.